docs/tutorial/frontend-tut.md

```diff
@@ -1,94 +1,117 @@
+2. We specify `sos`, a register. The `output` keyword means that we will populate `sos` in our program.
+3. The `map` operation gets the values of `avec` and raises each to the second power. We stash the result in a new array, `squares`. The number `1` denotes a *parallelism factor* of 1, meaning that the operation is performed sequentially. We will improve this shortly.
+4. The `reduce` operation walks over `squares` and accumulates the result into a register. The parallelism factor is again `1`.
 
 
-## Run a MrXL Program
+## Running our example
 
-Once we have [installed the mrxl command line tool][mrxldocs], we can run MrXL programs using [`fud`][fud].
+Let's run this program.
 
-To provide MrXL program with input values, we use fud's [JSON][json]-based [data format][fud-data].
-Let's try to run this program, which has a parallelism factor of two:
+To begin, [install the MrXL command line tool][mrxldocs-install].
+(TK: I'd kinda like to inline those instructions here, so that there is less jumping around between files.
+I want folks to just install and get back here so we can interpret together, but there's a risk that folks will just begin to follow the other guide and go all the way to Verilog there.
+Also, the MrXL tutorial has a different running example.
+Thoughts on the inline?)
 
+Now change directories to `calyx/frontends/mrxl` and run:
 ```
-{{#include ../../frontends/mrxl/test/add.mrxl}}
+mrxl test/sos.mrxl --data test/sos.mrxl.data --interpret
 ```
-In order to take advantage of the parallelism in the program, the MrXL compiler automatically partitions the input memory `foo` into two different *physical banks*: `foo_b0` and `foo_b1`.
-Therefore, we split up our logical `foo` input of `[1, 2, 3, 4]` into `[1,2]` and `[3,4]`:
+
+Why `42`? Because we populated `avec` with:
 ```json
-{{#include ../../frontends/mrxl/test/add.mrxl.data:2:23}}
+{{#include ../../frontends/mrxl/test/sos.mrxl.data}}
 ```
-Our complete data file similarly splits up the input for `baz`.
+and $0^2 + 1^2 + 4^2 + 5^2 = 42$.
+
 
-Run the program with the complete data by typing:
+## Compiling our example into Calyx
+
+Above, we merely _interpreted_ MrXL code in software using a simple, pre-written interpreter implemented in Python.
+Our goal in this tutorial is to build a compiler from MrXL to hardware by translating it to the Calyx IL.
+The Calyx IL code generated by compiling this program looks more like:
+```
+{{#include ./sos.calyx}}
+```
 
+Generate it for yourself! Run:
 ```
-fud exec frontends/mrxl/test/add.mrxl \
-    --from mrxl \
-    --to vcd -s verilog.data frontends/mrxl/test/add.mrxl.data
+mrxl test/sos.mrxl
 ```
 
-## Compiling MrXL to Calyx
+# Compiling MrXL into Calyx
+
+Any frontend can support new commands that are relevant to the domain of interest, so long as those commands can themselves be compiled into Calyx.
```

Comment:
The first line doesn't make too much sense to me. Is it trying to say that "new DSLs are useful for generating hardware"?

Reply:
I sorta wanted to say "You can design whatever DSL fanciness you want, but there is a price to pay for each fancy feature: you need to show your compiler how to take that feature to Calyx-land."